### DISCUSSION TITLE
godbolt compiler uses stack pointer register.

### DISCUSSION THREAD

**Discussion by nicholas132000** [NUS-CG3207/labs#23] [Lab 4] [answered]
Hello everybody, after i used godbolt compiler risc-v clang on a simple code as such, i noticed that it uses the SP register. Using the rars simulation, i see that SP is initialised to a value of 0x00003ffc. Under my test wrapper, the result i was expecting doesnt show up. Can i ask if i am supposed to intiialise my RegBank[2] (aka SP) in RegFile.v to 0x00003ffc? 
![image](https://github.com/user-attachments/assets/0ecc7d57-9277-4490-92b8-a9341a7859db)
![image](https://github.com/user-attachments/assets/e5afd92e-8078-4dc1-a04a-a5cc0f2837b8)

**Comment by r-panicker** [comment, marked as answer]
As mentioned here - https://nus-cg3207.github.io/lab-manuals/lab4/Using_Compiled_Code.html, stack pointer has to be properly initialised. 
Initializing in hardware is perfectly ok.

Inline assembly as shown in the example .c file (that you can find in the repo folder linked in the page above) might be an easier option. 

Either way, don't intialize it to 0x00003ffc unless your data memory address range includes 0x00003ffc. Else, the stack pointer points to a location that doesn't exist in your hardware. Assuming your data memory base is 0x2000 (which is the case for the third option for memory configuration in RARS), your data memory will need to be at least 0x1ffc in size, which means DMEM_DEPTH_BITS = 13. 

On a slightly different note, chances are that your compiler settings isn't -Os looking at the generated code. Optimised compiler output typically uses much less stack than unoptimized.

**Reply by nicholas132000** [reply]
got it prof! thanks:)
